grpc-client-server/prober_server/main_test.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func dialer() func(context.Context, string) (net.Conn, error) {
+	listener := bufconn.Listen(1024 * 1024)
+
+	server := grpc.NewServer()
+
+	pb.RegisterProberServer(server, &Server{})
+
+	go func() {
+		if err := server.Serve(listener); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	return func(context.Context, string) (net.Conn, error) {
+		return listener.Dial()
+	}
+}
+
+func TestDoProbes(t *testing.T) {
+	tests := map[string]struct {
+		req       *pb.ProbeRequest
+		err       error
+		failedReq int32
+	}{
+		"simple": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "https://www.google.com",
+				NumberOfRequests: 1,
+			},
+			failedReq: 0,
+		},
+		"error": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "http://musterror:8080",
+				NumberOfRequests: 3,
+			},
+			failedReq: 3,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer()), grpc.WithTransportCredentials(insecure.NewCredentials()))
+			require.NoError(t, err)
+			defer conn.Close()
+
+			client := pb.NewProberClient(conn)
+
+			resp, err := client.DoProbes(context.Background(), tc.req)
+
+			if err == nil {
+				require.NotNil(t, resp.TtfbAverageResponseTime)
+			}
+
+			require.Equal(t, tc.failedReq, resp.FailedRequests)
+		})
+	}
+
+	t.Run("response time", func(t *testing.T) {
+		conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer()), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		require.NoError(t, err)
+		defer conn.Close()
+
+		client := pb.NewProberClient(conn)
+
+		timeSince = func(t time.Time) time.Duration {
+			return 2 * time.Second
+		}
+
+		resp, err := client.DoProbes(context.Background(), &pb.ProbeRequest{
+			Endpoint:         "https://www.google.com",
+			NumberOfRequests: 1,
+		})
+
+		require.NoError(t, err)
+		require.NotNil(t, resp.TtfbAverageResponseTime)
+		require.NotNil(t, resp.TtlbAverageResponseTime)
+		require.Greater(t, resp.TtfbAverageResponseTime.AsDuration(), time.Duration(0))
+		require.Equal(t, 2*time.Second, resp.TtlbAverageResponseTime.AsDuration())
```

Comment:
One of these assertions is really specific, and the other really vague - what makes it hard to be specific in both?

Reply:
The challenge was to do with 2 different timing methods, 1 for TTFB and 1 for TTLB and that they are in 2 different places. 

I didn't want to measure TTLB inside the roundtripper as it will read everything from the body and return empty body. For our use case this would be fine, but I still think it's bad practice to be modified something that's expected to not be modified. 

We could of course make a copy of the body and return that, but this would use up unwanted resources. 

I have now implemented a solution for precise testing, which required using `timeSince` mock for both the server and Roundtripper.